kafka-connect-s3/src/main/java/io/confluent/connect/s3/TopicPartitionWriter.java

```diff
@@ -249,43 +266,56 @@ private boolean checkRotationOrAppend(
       String encodedPartition,
       long now
   ) {
+    // rotateOnTime check must go before writeRecord.
+    if (rotateOnTime(encodedPartition, currentTimestamp, now)) {
+      setNextScheduledRotation();
+      nextState();
+      return true;
+    }
+
+    // Conflict 1:
+    // If shouldChangeSchema is true, a rotation must be performed before writeRecord
+    // to avoid mismatching schemas in the writer.
+    // Additionally, this check must happen ahead of writeRecord,
+    // otherwise a SchemaProjectionException may be missed.
```

Comment:
It's sufficient to assume that faults are independent of schema mismatches. If both happen in the user's code then it's fine for the rotation to happen. `null` keys and empty headers are not allowed when key and header export is enabled. 
```suggestion
```